### PR TITLE
[Optimization](index) Optimization for no need to read raw data for index column that only in where clause

### PR DESCRIPTION
**Pull request by Tanya-W**
# Proposed changes

Issue Number: close #xxx

## Problem summary
At the storage layer, the raw data of the index column will still be read after apply index(bitmap_index or inverted_index), although the index column is not in the result column returned by the query, that will generate more performance overhead on seek and read data.

In addition, when there are multi-table join query, there will be many in or not_in predicate of runtime filter pushed down to the storage layer. According to our test,  if apply those predicates by inverted index, the performance will be degraded because there are many conditions in in_predicate. Therefore, the inverted index not apply on in or not_in predicate which is produced by runtime_filter.

Based on that situation, this pr will do:
1. reduce overhead on seek and read data for index column that only in where clause, optimization for query sql like:
```
sql 1: SELECT timestamp FROM tb WHERE log MATCH 'error';
sql 2: SELECT timestamp FROM tb WHERE log MATCH 'error' ORDER BY timestamp LIMIT 2;
sql 3: SELECT timestamp FROM tb WHERE log MATCH 'error' AND status = 404;
sql 4: SELECT timestamp FROM tb WHERE log MATCH 'error' AND status = 404 ORDER BY timestamp DESC LIMIT 10;
sql 5: SELECT count() FROM tb WHERE log MATCH 'error';
```
column `log` and column `status` is inverted index or bitmap index, above sqls only need seek and read data of column `timestamp`

2. not apply inverted index on in or not_in predicate which is produced by runtime_filter.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
4. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
5. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
6. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
7. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/olap/rowset/segment_v2/segment_iterator.cpp [hunk 1]
```diff
@@ -1562,5 +1633,83 @@ Status SegmentIterator::current_block_row_locations(std::vector<RowLocation>* bl
...
+bool SegmentIterator::_check_column_pred_all_push_down(ColumnPredicate* predicate,
+                                                       bool in_compound) {
+    if (_remaining_vconjunct_root == nullptr) {
+        return true;
+    }
+    std::unordered_map<std::string, std::vector<ColumnPredicateInfo>>
+            column_pred_in_remaining_vconjunct;
+    auto column_name = _schema.column(predicate->column_id())->name();
+    _column_predicate_info.reset(new ColumnPredicateInfo());
+    _find_pred_in_remaining_vconjunct_root(_remaining_vconjunct_root,
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/segment_iterator.cpp:1664]: should only get column_name related predicated

File: gensrc/thrift/PlanNodes.thrift [hunk 2]
```diff
@@ -574,6 +574,8 @@ struct TOlapScanNode {
   11: optional bool enable_unique_key_merge_on_write
   12: optional TPushAggOp push_down_agg_type_opt
   13: optional bool use_topn_opt
+  14: optional list<Exprs.TExpr> ordering_exprs
```
[the remark this answers is not in the document]
Comment (Tanya-W) [reply gensrc/thrift/PlanNodes.thrift:577]: unused, I will remove it

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.58 seconds
 stream load tsv:          470 seconds loaded 74807831229 Bytes, about 151 MB/s
 stream load json:         36 seconds loaded 2358488459 Bytes, about 62 MB/s
 stream load orc:          68 seconds loaded 1101869774 Bytes, about 15 MB/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230209142416_clickbench_pr_93285.html

File: fe/fe-core/src/main/java/org/apache/doris/planner/OriginalPlanner.java [hunk 3]
```diff
@@ -436,6 +441,60 @@ private void pushSortToOlapScan() {
...
+                        }
+                    }
+                }
+                // OlapScanNode is the last node.
+                // So, just get the two node and check if they are SortNode and OlapScan.
+                parent = node;
+                node = node.getChildren().get(0);
+            }
+
+            if (parent instanceof SortNode) {
```
[the remark this answers is not in the document]
Comment (Tanya-W) [reply fe/fe-core/src/main/java/org/apache/doris/planner/OriginalPlanner.java:478]: GROUP BY columns must in SELECT output columns list

File: be/src/olap/rowset/segment_v2/segment_iterator.cpp (hunk 1 above)
Comment (Tanya-W) [reply be/src/olap/rowset/segment_v2/segment_iterator.cpp:1664]: you are right, here is a bug, I will fix it.

File: be/src/olap/rowset/segment_v2/segment_iterator.cpp [hunk 4]
```diff
@@ -1562,5 +1633,83 @@ Status SegmentIterator::current_block_row_locations(std::vector<RowLocation>* bl
...
+ *      column C : `C like '%ye%'` not pushed down, `C > 'aa'` pushed down, only `C > 'aa'` applyed by index,
+ *                  call _check_column_pred_all_push_down will return false.
+*/
+bool SegmentIterator::_check_column_pred_all_push_down(ColumnPredicate* predicate,
+                                                       bool in_compound) {
+    if (_remaining_vconjunct_root == nullptr) {
+        return true;
+    }
+    std::unordered_map<std::string, std::vector<ColumnPredicateInfo>>
+            column_pred_in_remaining_vconjunct;
```
[the remark this answers is not in the document]
Comment (Tanya-W) [reply be/src/olap/rowset/segment_v2/segment_iterator.cpp:1661]: yes, you are right, I will update this logic

File: fe/fe-core/src/main/java/org/apache/doris/planner/OriginalPlanner.java [hunk 5]
```diff
@@ -436,6 +441,60 @@ private void pushSortToOlapScan() {
         }
     }
 
+    /**
+     * outputColumnUniqueIds contain columns in OrderByExprs and outputExprs
+     * push output column unique id set to olap scan.
```
[the remark this answers is not in the document]
Comment (Tanya-W) [reply fe/fe-core/src/main/java/org/apache/doris/planner/OriginalPlanner.java:446]: updated

File: be/src/olap/reader.h [hunk 6]
```diff
@@ -102,6 +102,7 @@ class TabletReader {
 
         std::vector<RowsetReaderSharedPtr> rs_readers;
         std::vector<uint32_t> return_columns;
+        std::set<int32_t> output_columns;
```
[the remark this answers is not in the document]
Comment (Tanya-W) [reply be/src/olap/reader.h:105]: updated

**Review by xiaokang — APPROVED** [review]
LGTM

**Review by dataroaring — APPROVED** [review]
LGTM

**Pull request merged**